// @datadog/datadog-ci/dist/commands/sarif/renderer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.renderFilesNotFound = exports.renderCommandInfo = exports.renderUpload = exports.renderDryRunUpload = exports.renderSuccessfulCommand = exports.renderRetriedUpload = exports.renderFailedUpload = exports.renderInvalidFile = void 0;
const path_1 = __importDefault(require("path"));
const chalk_1 = __importDefault(require("chalk"));
const ICONS = {
    FAILED: '❌',
    SUCCESS: '✅',
    WARNING: '⚠️',
    INFO: 'ℹ️',
};
const renderInvalidFile = (sarifReport, errorMessage) => {
    let fullStr = '';
    const reportPath = `[${chalk_1.default.bold.dim(sarifReport)}]`;
    fullStr += chalk_1.default.red(`${ICONS.FAILED} Invalid SARIF report file ${reportPath}.\n`);
    fullStr += chalk_1.default.red(`The report is not a valid JSON or is not compliant with the SARIF json schema v2.1.0.\n`);
    fullStr += chalk_1.default.red(`Errors: ${errorMessage}\n`);
    return fullStr;
};
exports.renderInvalidFile = renderInvalidFile;
const renderFailedUpload = (sarifReport, errorMessage) => {
    const reportPath = `[${chalk_1.default.bold.dim(sarifReport.reportPath)}]`;
    return chalk_1.default.red(`${ICONS.FAILED} Failed upload SARIF report file ${reportPath}: ${errorMessage}\n`);
};
exports.renderFailedUpload = renderFailedUpload;
const renderRetriedUpload = (sarifReport, errorMessage, attempt) => {
    const sarifReportPath = `[${chalk_1.default.bold.dim(sarifReport.reportPath)}]`;
    return chalk_1.default.yellow(`[attempt ${attempt}] Retrying SARIF report upload ${sarifReportPath}: ${errorMessage}\n`);
};
exports.renderRetriedUpload = renderRetriedUpload;
const renderSuccessfulCommand = (fileCount, duration, spanTags, service, env) => {
    let fullStr = '';
    fullStr += chalk_1.default.green(`${ICONS.SUCCESS} Uploaded ${fileCount} files in ${duration} seconds.\n`);
    fullStr += chalk_1.default.green('=================================================================================================\n');
    return fullStr;
};
exports.renderSuccessfulCommand = renderSuccessfulCommand;
const renderDryRunUpload = (payload) => `[DRYRUN] ${(0, exports.renderUpload)(payload)}`;
exports.renderDryRunUpload = renderDryRunUpload;
const renderUpload = (payload) => `Uploading SARIF report in ${payload.reportPath}\n`;
exports.renderUpload = renderUpload;
const renderCommandInfo = (basePaths, service, concurrency, dryRun, noVerify) => {
    let fullStr = '';
    if (dryRun) {
        fullStr += chalk_1.default.yellow(`${ICONS.WARNING} DRY-RUN MODE ENABLED. WILL NOT UPLOAD SARIF REPORT\n`);
    }
    if (noVerify) {
        fullStr += chalk_1.default.yellow(`${ICONS.INFO} --no-verify enabled. The reports will be uploaded without client validation.\n`);
    }
    fullStr += chalk_1.default.green(`Starting upload with concurrency ${concurrency}. \n`);
    if (basePaths.length === 1 && !!path_1.default.extname(basePaths[0])) {
        fullStr += chalk_1.default.green(`Will upload SARIF report file ${basePaths[0]}\n`);
    }
    else {
        fullStr += chalk_1.default.green(`Will look for SARIF report files in ${basePaths.join(', ')}\n`);
    }
    fullStr += chalk_1.default.green(`service: ${service}\n`);
    return fullStr;
};
exports.renderCommandInfo = renderCommandInfo;
const renderFilesNotFound = (basePaths, service) => {
    let fullStr = '';
    const paths = basePaths.length === 1 && !!path_1.default.extname(basePaths[0]) ? basePaths[0] : basePaths.join(', ');
    fullStr += chalk_1.default.yellow(`${ICONS.WARNING} Cannot find valid SARIF report files to upload in ${paths} for service ${service}.\n`);
    fullStr += chalk_1.default.yellow(`Check the files exist and are valid.\n`);
    return fullStr;
};
exports.renderFilesNotFound = renderFilesNotFound;
