// @datadog/datadog-ci/dist/commands/sourcemaps/interfaces.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Sourcemap = void 0;
const fs_1 = __importDefault(require("fs"));
class Sourcemap {
    constructor(minifiedFilePath, minifiedUrl, sourcemapPath, relativePath, minifiedPathPrefix) {
        this.minifiedFilePath = minifiedFilePath;
        this.minifiedPathPrefix = minifiedPathPrefix;
        this.minifiedUrl = minifiedUrl;
        this.relativePath = relativePath;
        this.sourcemapPath = sourcemapPath;
    }
    addRepositoryData(gitData) {
        this.gitData = gitData;
    }
    asMultipartPayload(cliVersion, service, version, projectPath) {
        const content = new Map([
            ['event', this.getMetadataPayload(cliVersion, service, version, projectPath)],
            ['source_map', { value: fs_1.default.createReadStream(this.sourcemapPath), options: { filename: 'source_map' } }],
            ['minified_file', { value: fs_1.default.createReadStream(this.minifiedFilePath), options: { filename: 'minified_file' } }],
        ]);
        if (this.gitData !== undefined && this.gitData.gitRepositoryPayload !== undefined) {
            content.set('repository', {
                options: {
                    contentType: 'application/json',
                    filename: 'repository',
                },
                value: this.gitData.gitRepositoryPayload,
            });
        }
        return {
            content,
        };
    }
    getMetadataPayload(cliVersion, service, version, projectPath) {
        const metadata = {
            cli_version: cliVersion,
            minified_url: this.minifiedUrl,
            project_path: projectPath,
            service,
            type: 'js_sourcemap',
            version,
        };
        if (this.gitData !== undefined) {
            metadata.git_repository_url = this.gitData.gitRepositoryURL;
            metadata.git_commit_sha = this.gitData.gitCommitSha;
        }
        return {
            options: {
                contentType: 'application/json',
                filename: 'event',
            },
            value: JSON.stringify(metadata),
        };
    }
}
exports.Sourcemap = Sourcemap;
