// auth0-deploy-cli/lib/tools/auth0/handlers/guardianPhoneFactorSelectedProvider.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.schema = undefined;

var _default = require("./default");

var _default2 = _interopRequireDefault(_default);

var _constants = require("../../constants");

var _constants2 = _interopRequireDefault(_constants);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const schema = exports.schema = {
  type: 'object',
  properties: {
    provider: {
      type: 'string',
      enum: _constants2.default.GUARDIAN_PHONE_PROVIDERS
    }
  },
  additionalProperties: false
};

const isFeatureUnavailableError = err => {
  if (err.statusCode === 404) {
    return true;
  }

  if (err.statusCode === 403 && err.originalError && err.originalError.response && err.originalError.response.body && err.originalError.response.body.errorCode === 'hooks_not_allowed') {
    return true;
  }

  return false;
};

class GuardianPhoneSelectedProviderHandler extends _default2.default {
  constructor(options) {
    super({ ...options,
      type: 'guardianPhoneFactorSelectedProvider'
    });
  }

  async getType() {
    if (!this.client.guardian || typeof this.client.guardian.getPhoneFactorSelectedProvider !== 'function') {
      return {};
    }

    if (this.existing) return this.existing;

    try {
      this.existing = await this.client.guardian.getPhoneFactorSelectedProvider();
    } catch (e) {
      if (isFeatureUnavailableError(e)) {
        return {};
      }

      throw e;
    }

    return this.existing;
  }

  async processChanges(assets) {
    const {
      guardianPhoneFactorSelectedProvider
    } = assets;
    if (!guardianPhoneFactorSelectedProvider || !guardianPhoneFactorSelectedProvider.provider) return;
    const params = {};
    const data = guardianPhoneFactorSelectedProvider;
    await this.client.guardian.updatePhoneFactorSelectedProvider(params, data);
    this.updated += 1;
    this.didUpdate(guardianPhoneFactorSelectedProvider);
  }

}

exports.default = GuardianPhoneSelectedProviderHandler;
