// auth0-deploy-cli/lib/tools/auth0/handlers/guardianFactors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.schema = undefined;

var _default = require("./default");

var _default2 = _interopRequireDefault(_default);

var _constants = require("../../constants");

var _constants2 = _interopRequireDefault(_constants);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const schema = exports.schema = {
  type: 'array',
  items: {
    type: 'object',
    properties: {
      name: {
        type: 'string',
        enum: _constants2.default.GUARDIAN_FACTORS
      }
    },
    required: ['name']
  }
};

class GuardianFactorsHandler extends _default2.default {
  constructor(options) {
    super({ ...options,
      type: 'guardianFactors',
      id: 'name'
    });
  }

  async getType() {
    if (this.existing) return this.existing;
    this.existing = await this.client.guardian.getFactors();
    return this.existing;
  }

  async processChanges(assets) {
    const {
      guardianFactors
    } = assets;
    if (!guardianFactors || !guardianFactors.length) return;
    await Promise.all(guardianFactors.map(async factor => {
      const data = { ...factor
      };
      const params = {
        name: factor.name
      };
      delete data.name;
      await this.client.guardian.updateFactor(params, data);
      this.didUpdate(params);
      this.updated += 1;
    }));
  }

}

exports.default = GuardianFactorsHandler;
