// auth0-deploy-cli/lib/tools/auth0/handlers/guardianPolicies.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.schema = undefined;

var _default = require("./default");

var _default2 = _interopRequireDefault(_default);

var _constants = require("../../constants");

var _constants2 = _interopRequireDefault(_constants);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const schema = exports.schema = {
  type: 'object',
  properties: {
    policies: {
      type: 'array',
      items: {
        type: 'string',
        enum: _constants2.default.GUARDIAN_POLICIES
      }
    }
  },
  additionalProperties: false
};

class GuardianPoliciesHandler extends _default2.default {
  constructor(options) {
    super({ ...options,
      type: 'guardianPolicies'
    });
  }

  async getType() {
    if (!this.client.guardian || typeof this.client.guardian.getPolicies !== 'function') {
      return {};
    }

    if (this.existing) return this.existing;
    const policies = await this.client.guardian.getPolicies();
    this.existing = {
      policies
    };
    return this.existing;
  }

  async processChanges(assets) {
    const {
      guardianPolicies
    } = assets;
    if (!guardianPolicies || !guardianPolicies.policies) return;
    const params = {};
    const data = guardianPolicies.policies;
    await this.client.guardian.updatePolicies(params, data);
    this.updated += 1;
    this.didUpdate(guardianPolicies);
  }

}

exports.default = GuardianPoliciesHandler;
