// auth0-deploy-cli/lib/tools/auth0/handlers/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.organizations = exports.triggers = exports.actions = exports.migrations = exports.prompts = exports.branding = exports.roles = exports.guardianPhoneFactorMessageTypes = exports.guardianPhoneFactorSelectedProvider = exports.guardianPolicies = exports.guardianFactorTemplates = exports.guardianFactorProviders = exports.guardianFactors = exports.clientGrants = exports.emailTemplates = exports.emailProvider = exports.tenant = exports.connections = exports.databases = exports.clients = exports.resourceServers = exports.pages = exports.hooks = exports.rulesConfigs = exports.rules = undefined;

var _rules = require("./rules");

var rules = _interopRequireWildcard(_rules);

var _rulesConfigs = require("./rulesConfigs");

var rulesConfigs = _interopRequireWildcard(_rulesConfigs);

var _hooks = require("./hooks");

var hooks = _interopRequireWildcard(_hooks);

var _pages = require("./pages");

var pages = _interopRequireWildcard(_pages);

var _resourceServers = require("./resourceServers");

var resourceServers = _interopRequireWildcard(_resourceServers);

var _databases = require("./databases");

var databases = _interopRequireWildcard(_databases);

var _connections = require("./connections");

var connections = _interopRequireWildcard(_connections);

var _clients = require("./clients");

var clients = _interopRequireWildcard(_clients);

var _tenant = require("./tenant");

var tenant = _interopRequireWildcard(_tenant);

var _emailProvider = require("./emailProvider");

var emailProvider = _interopRequireWildcard(_emailProvider);

var _emailTemplates = require("./emailTemplates");

var emailTemplates = _interopRequireWildcard(_emailTemplates);

var _clientGrants = require("./clientGrants");

var clientGrants = _interopRequireWildcard(_clientGrants);

var _guardianFactors = require("./guardianFactors");

var guardianFactors = _interopRequireWildcard(_guardianFactors);

var _guardianFactorProviders = require("./guardianFactorProviders");

var guardianFactorProviders = _interopRequireWildcard(_guardianFactorProviders);

var _guardianFactorTemplates = require("./guardianFactorTemplates");

var guardianFactorTemplates = _interopRequireWildcard(_guardianFactorTemplates);

var _guardianPolicies = require("./guardianPolicies");

var guardianPolicies = _interopRequireWildcard(_guardianPolicies);

var _guardianPhoneFactorSelectedProvider = require("./guardianPhoneFactorSelectedProvider");

var guardianPhoneFactorSelectedProvider = _interopRequireWildcard(_guardianPhoneFactorSelectedProvider);

var _guardianPhoneFactorMessageTypes = require("./guardianPhoneFactorMessageTypes");

var guardianPhoneFactorMessageTypes = _interopRequireWildcard(_guardianPhoneFactorMessageTypes);

var _roles = require("./roles");

var roles = _interopRequireWildcard(_roles);

var _branding = require("./branding");

var branding = _interopRequireWildcard(_branding);

var _prompts = require("./prompts");

var prompts = _interopRequireWildcard(_prompts);

var _migrations = require("./migrations");

var migrations = _interopRequireWildcard(_migrations);

var _actions = require("./actions");

var actions = _interopRequireWildcard(_actions);

var _triggers = require("./triggers");

var triggers = _interopRequireWildcard(_triggers);

var _organizations = require("./organizations");

var organizations = _interopRequireWildcard(_organizations);

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

exports.rules = rules;
exports.rulesConfigs = rulesConfigs;
exports.hooks = hooks;
exports.pages = pages;
exports.resourceServers = resourceServers;
exports.clients = clients;
exports.databases = databases;
exports.connections = connections;
exports.tenant = tenant;
exports.emailProvider = emailProvider;
exports.emailTemplates = emailTemplates;
exports.clientGrants = clientGrants;
exports.guardianFactors = guardianFactors;
exports.guardianFactorProviders = guardianFactorProviders;
exports.guardianFactorTemplates = guardianFactorTemplates;
exports.guardianPolicies = guardianPolicies;
exports.guardianPhoneFactorSelectedProvider = guardianPhoneFactorSelectedProvider;
exports.guardianPhoneFactorMessageTypes = guardianPhoneFactorMessageTypes;
exports.roles = roles;
exports.branding = branding;
exports.prompts = prompts;
exports.migrations = migrations;
exports.actions = actions;
exports.triggers = triggers;
exports.organizations = organizations;
