// auth0-deploy-cli/lib/tools/auth0/handlers/rulesConfigs.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.schema = undefined;

var _default = require("./default");

var _default2 = _interopRequireDefault(_default);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const schema = exports.schema = {
  type: 'array',
  items: {
    type: 'object',
    properties: {
      key: {
        type: 'string',
        pattern: '^[A-Za-z0-9_-]*$'
      },
      value: {
        type: 'string'
      }
    },
    required: ['key', 'value']
  },
  additionalProperties: false
};

class RulesConfigsHandler extends _default2.default {
  constructor(options) {
    super({ ...options,
      type: 'rulesConfigs',
      id: 'key',
      functions: {
        update: 'set'
      }
    });
  }

  async getType() {
    return this.client.rulesConfigs.getAll();
  }

  objString(item) {
    return super.objString({
      key: item.key
    });
  }

  async calcChanges(assets) {
    const {
      rulesConfigs
    } = assets;
    if (!rulesConfigs || !rulesConfigs.length) return {};
    return {
      del: [],
      update: rulesConfigs,
      create: []
    };
  }

}

exports.default = RulesConfigsHandler;
