// auth0-deploy-cli/lib/context/directory/handlers/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _pages = require("./pages");

var _pages2 = _interopRequireDefault(_pages);

var _rules = require("./rules");

var _rules2 = _interopRequireDefault(_rules);

var _hooks = require("./hooks");

var _hooks2 = _interopRequireDefault(_hooks);

var _clients = require("./clients");

var _clients2 = _interopRequireDefault(_clients);

var _tenant = require("./tenant");

var _tenant2 = _interopRequireDefault(_tenant);

var _emailProvider = require("./emailProvider");

var _emailProvider2 = _interopRequireDefault(_emailProvider);

var _connections = require("./connections");

var _connections2 = _interopRequireDefault(_connections);

var _databases = require("./databases");

var _databases2 = _interopRequireDefault(_databases);

var _emailTemplates = require("./emailTemplates");

var _emailTemplates2 = _interopRequireDefault(_emailTemplates);

var _clientGrants = require("./clientGrants");

var _clientGrants2 = _interopRequireDefault(_clientGrants);

var _rulesConfigs = require("./rulesConfigs");

var _rulesConfigs2 = _interopRequireDefault(_rulesConfigs);

var _resourceServers = require("./resourceServers");

var _resourceServers2 = _interopRequireDefault(_resourceServers);

var _guardianFactors = require("./guardianFactors");

var _guardianFactors2 = _interopRequireDefault(_guardianFactors);

var _guardianFactorProviders = require("./guardianFactorProviders");

var _guardianFactorProviders2 = _interopRequireDefault(_guardianFactorProviders);

var _guardianFactorTemplates = require("./guardianFactorTemplates");

var _guardianFactorTemplates2 = _interopRequireDefault(_guardianFactorTemplates);

var _guardianPhoneFactorMessageTypes = require("./guardianPhoneFactorMessageTypes");

var _guardianPhoneFactorMessageTypes2 = _interopRequireDefault(_guardianPhoneFactorMessageTypes);

var _guardianPhoneFactorSelectedProvider = require("./guardianPhoneFactorSelectedProvider");

var _guardianPhoneFactorSelectedProvider2 = _interopRequireDefault(_guardianPhoneFactorSelectedProvider);

var _guardianPolicies = require("./guardianPolicies");

var _guardianPolicies2 = _interopRequireDefault(_guardianPolicies);

var _roles = require("./roles");

var _roles2 = _interopRequireDefault(_roles);

var _migrations = require("./migrations");

var _migrations2 = _interopRequireDefault(_migrations);

var _actions = require("./actions");

var _actions2 = _interopRequireDefault(_actions);

var _organizations = require("./organizations");

var _organizations2 = _interopRequireDefault(_organizations);

var _triggers = require("./triggers");

var _triggers2 = _interopRequireDefault(_triggers);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  rules: _rules2.default,
  rulesConfigs: _rulesConfigs2.default,
  hooks: _hooks2.default,
  pages: _pages2.default,
  databases: _databases2.default,
  clientGrants: _clientGrants2.default,
  resourceServers: _resourceServers2.default,
  clients: _clients2.default,
  connections: _connections2.default,
  tenant: _tenant2.default,
  emailProvider: _emailProvider2.default,
  emailTemplates: _emailTemplates2.default,
  guardianFactors: _guardianFactors2.default,
  guardianFactorProviders: _guardianFactorProviders2.default,
  guardianFactorTemplates: _guardianFactorTemplates2.default,
  migrations: _migrations2.default,
  guardianPhoneFactorMessageTypes: _guardianPhoneFactorMessageTypes2.default,
  guardianPhoneFactorSelectedProvider: _guardianPhoneFactorSelectedProvider2.default,
  guardianPolicies: _guardianPolicies2.default,
  roles: _roles2.default,
  actions: _actions2.default,
  organizations: _organizations2.default,
  triggers: _triggers2.default
};
