// @datadog/datadog-ci/dist/commands/react-native/interfaces.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RN_SUPPORTED_PLATFORMS = exports.RNSourcemap = void 0;
const fs_1 = __importDefault(require("fs"));
class RNSourcemap {
    constructor(bundleName, sourcemapPath) {
        this.removeSourcesContentFromSourceMap = () => {
            const newSourcemapFilePath = `${this.sourcemapPath}.no-sources-content`;
            const data = fs_1.default.readFileSync(this.sourcemapPath, 'utf8');
            const sourcemap = JSON.parse(data);
            delete sourcemap.sourcesContent;
            fs_1.default.writeFileSync(newSourcemapFilePath, JSON.stringify(sourcemap), 'utf8');
            this.sourcemapPath = newSourcemapFilePath;
        };
        this.bundleName = bundleName;
        this.sourcemapPath = sourcemapPath;
    }
    addRepositoryData(gitData) {
        this.gitData = gitData;
    }
    asMultipartPayload(cliVersion, service, version, projectPath, platform, build) {
        const content = new Map([
            ['event', this.getMetadataPayload(cliVersion, service, version, projectPath, platform, build)],
            ['source_map', { value: fs_1.default.createReadStream(this.sourcemapPath), options: { filename: 'source_map' } }],
        ]);
        if (this.gitData !== undefined && this.gitData.gitRepositoryPayload !== undefined) {
            content.set('repository', {
                options: {
                    contentType: 'application/json',
                    filename: 'repository',
                },
                value: this.gitData.gitRepositoryPayload,
            });
        }
        return {
            content,
        };
    }
    getMetadataPayload(cliVersion, service, version, projectPath, platform, build) {
        const metadata = {
            build_number: build,
            bundle_name: this.bundleName,
            cli_version: cliVersion,
            platform,
            project_path: projectPath,
            service,
            type: 'react_native_sourcemap',
            version,
        };
        if (this.gitData !== undefined) {
            metadata.git_repository_url = this.gitData.gitRepositoryURL;
            metadata.git_commit_sha = this.gitData.gitCommitSha;
        }
        return {
            options: {
                contentType: 'application/json',
                filename: 'event',
            },
            value: JSON.stringify(metadata),
        };
    }
}
exports.RNSourcemap = RNSourcemap;
exports.RN_SUPPORTED_PLATFORMS = ['ios', 'android'];
