// @datadog/datadog-ci/dist/commands/dsyms/interfaces.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CompressedDsym = void 0;
const fs_1 = __importDefault(require("fs"));
class CompressedDsym {
    constructor(archivePath, dsym) {
        this.archivePath = archivePath;
        this.dsym = dsym;
    }
    asMultipartPayload() {
        const content = new Map([
            ['symbols_archive', { value: fs_1.default.createReadStream(this.archivePath), options: { filename: 'ios_symbols_archive' } }],
            ['event', this.getMetadataPayload()],
        ]);
        return {
            content,
        };
    }
    getMetadataPayload() {
        const concatUUIDs = this.dsym.slices.map((slice) => slice.uuid).join();
        return {
            options: {
                contentType: 'application/json',
                filename: 'event',
            },
            value: JSON.stringify({
                type: 'ios_symbols',
                uuids: concatUUIDs,
            }),
        };
    }
}
exports.CompressedDsym = CompressedDsym;
