// auth0-deploy-cli/lib/commands/export.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = deploy;

var _path = require("path");

var _path2 = _interopRequireDefault(_path);

var _nconf = require("nconf");

var _nconf2 = _interopRequireDefault(_nconf);

var _mkdirp = require("mkdirp");

var _mkdirp2 = _interopRequireDefault(_mkdirp);

var _logger = require("../logger");

var _logger2 = _interopRequireDefault(_logger);

var _utils = require("../utils");

var _context = require("../context");

var _context2 = _interopRequireDefault(_context);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

async function deploy(params) {
  const {
    output_folder: outputFolder,
    base_path: basePath,
    config_file: configFile,
    config: configObj,
    export_ids: exportIds,
    secret
  } = params;

  _nconf2.default.env().use('memory');

  if (configFile) {
    _nconf2.default.file(configFile);
  }

  const overrides = {
    AUTH0_INPUT_FILE: outputFolder,
    AUTH0_BASE_PATH: basePath,
    AUTH0_CONFIG_FILE: configFile,
    ...(configObj || {})
  };

  if (secret) {
    overrides.AUTH0_CLIENT_SECRET = secret;
  }

  if (exportIds) {
    overrides.AUTH0_EXPORT_IDENTIFIERS = exportIds;
  }

  if (!(0, _utils.isDirectory)(outputFolder)) {
    _logger2.default.info(`Creating ${outputFolder}`);

    _mkdirp2.default.sync(outputFolder);
  }

  if (params.format === 'yaml') {
    overrides.AUTH0_INPUT_FILE = _path2.default.join(outputFolder, 'tenant.yaml');
  }

  _nconf2.default.overrides(overrides);

  const context = await (0, _context2.default)(_nconf2.default.get());
  await context.dump();

  _logger2.default.info('Export Successful');
}
