// auth0-deploy-cli/lib/commands/import.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = deploy;

var _nconf = require("nconf");

var _nconf2 = _interopRequireDefault(_nconf);

var _configFactory = require("../configFactory");

var _configFactory2 = _interopRequireDefault(_configFactory);

var _tools = require("../tools");

var _logger = require("../logger");

var _logger2 = _interopRequireDefault(_logger);

var _context = require("../context");

var _context2 = _interopRequireDefault(_context);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

async function deploy(params) {
  const {
    input_file: inputFile,
    base_path: basePath,
    config_file: configFile,
    config: configObj,
    env,
    secret
  } = params;

  _nconf2.default.env().use('memory');

  if (configFile) {
    _nconf2.default.file(configFile);
  }

  const overrides = {
    AUTH0_INPUT_FILE: inputFile,
    AUTH0_BASE_PATH: basePath,
    AUTH0_CONFIG_FILE: configFile,
    AUTH0_KEYWORD_REPLACE_MAPPINGS: {},
    ...(configObj || {})
  };

  if (secret) {
    overrides.AUTH0_CLIENT_SECRET = secret;
  }

  if (env) {
    const mappings = _nconf2.default.get('AUTH0_KEYWORD_REPLACE_MAPPINGS') || {};

    _nconf2.default.set('AUTH0_KEYWORD_REPLACE_MAPPINGS', Object.assign(mappings, process.env));
  }

  _nconf2.default.overrides(overrides);

  const context = await (0, _context2.default)(_nconf2.default.get());
  await context.load();
  const config = (0, _configFactory2.default)();
  config.setProvider(key => _nconf2.default.get(key));
  await (0, _tools.deploy)(context.assets, context.mgmtClient, config);

  _logger2.default.info('Import Successful');
}
